text/0000-hierarchic-anonymous-life-time.md

```diff
@@ -0,0 +1,136 @@
+- Feature Name: Hierarchic anonymous life-time
+- Start Date: 2020-06-24
+- RFC PR: [rust-lang/rfcs#0000](https://github.com/rust-lang/rfcs/pull/0000)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+
+New use of anonymous life-time '_ that implicitly added to current structure
+
+# Motivation
+
+Motivation is to simplify iterative development and improving refactoring of the code
+
+Sometimes during refactoring such code:
+```rust
+struct CompositeObject {
+    obj: SomeType,
+}
+
+struct BigObject {
+    composite_obj: CompositeObject,
+    count: i32,
+}
+
+struct Application {
+   big_obj: BigObject,
+}
+```
+
+developer decides to make obj of SomeType as reference in CompositeObject type:
+```rust
+struct CompositeObject<'a> {
+    obj: &'a SomeType,
+}
+
+struct BigObject<'a> {
+    composite_obj: CompositeObject<'a>,
+    count: i32,
+}
+
+struct Application<'a> {
+   big_obj: BigObject<'a>,
+}
+```
+Everywhere in composition hierarchy I need to write 'a ... most of the times it is just boilerplate code ...
+
+What if instead of writing manually we will specify reference fields with anonymous life-time:
+```rust
+struct CompositeObject {
+    obj: &'_ SomeType,
+}
+
+struct BigObject {
+    composite_obj: CompositeObject,
+    count: i32,
+}
+
+struct Application {
+   big_obj: BigObject,
+}
+```
+
+Code much simpler and more maintainable than fighting with named life-times in composite hierarchy
+
+Compiler underhood will generate the following code:
+```rust
+struct CompositeObject<'anon> {             // 'anon is implicitly added life-time
+    obj: &'anon SomeType,
+}
+
+struct BigObject<'anon> {                   // 'anon is implicitly added life-time
+    composite_obj: CompositeObject<'anon>,  // 'anon is implicitly used here
+    count: i32,
+}
+
+struct Application<'anon> {                 // 'anon is implicitly added life-time
+   big_obj: BigObject<'anon>,               // 'anon is implicitly used here
+}
```

Comment:
If the following code is generated, how will a reader knows that
```rust
struct BigObject {
    composite_obj: CompositeObject,
    count: i32,
}
```

have an implicit `CompositeObject<'anon>` by reading the documentation generated for `BigObject`?

Reply:
> 
> 
> If the following code is generated, how will a reader knows that
> 
> ```rust
> struct BigObject {
>     composite_obj: CompositeObject,
>     count: i32,
> }
> ```
> 
> have an implicit `CompositeObject<'anon>` by reading the documentation generated for `BigObject`?

@pickfire Actually thinking ... what explicit life-time gives you ? You just need to know how to initialize `CompositeObject` ... all other job should be done by `CompositeObject`, it is level of abstraction

Reply:
To me it has been quite helpful to see explicit lifetimes because then I know that there is a lifetime being tracked inside the structure. In that way, it is a very helpful kind of documentation. One that is also always up-to-date, which is not a common feature of documentation in general.

Reply:
> 
> 
> To me it has been quite helpful to see explicit lifetimes because then I know that there is a lifetime being tracked inside the structure. In that way, it is a very helpful kind of documentation. One that is also always up-to-date, which is not a common feature of documentation in general.

@felix91gr You answer question: "... it is a very helpful kind of documentation ...", it is mix of multiple stuffs in one ... it breakage of lot of pattern like `SOLID` - Single Responsibility

Lets not mix life-time with good documentation ... from my point you should only know how to initialize `CompositeObject` that is all

Responsibility of proper handling life-time is responsibility of `CompositeObject` and developer that written it

Bypassing life-time from "low-level" structures to "high-level" structures (logically) is violation of layers responsibilities

Each abstract layer souls be responsible for its stuff

Reply:
> it breakage of lot of pattern like SOLID - Single Responsibility

Those patterns were created as heuristics for Object-Oriented Programming, not as tips for general design.

In this case, having two responsibilities is good - it puts the information in exactly one place, instead of two. Keeping the documentation of what this lifetime is affecting becomes trivial once you have to keep it written, doesn't it?

Reply:
> 
> 
> > it breakage of lot of pattern like SOLID - Single Responsibility
> 
> Those patterns were created as heuristics for Object-Oriented Programming, not as tips for general design.
> 
> In this case, having two responsibilities is good - it puts the information in exactly one place, instead of two. Keeping the documentation of what this lifetime is affecting becomes trivial once you have to keep it written, doesn't it?

No, those patterns could be applied to system that want to be designed good ;)

And if you want good documentation it is possible to extend rustdoc in way it read information form AST (in AST implicit life-times are available) and then generate good documentation

Reply:
> Responsibility of proper handling life-time is responsibility of `CompositeObject` and developer that written it
>
> Bypassing life-time from "low-level" structures to "high-level" structures (logically) is violation of layers responsibilities

Actually, by storing references in `CompositeObject`, you are *explicitly* delegating the responsibility of managing the lifetime of the pointee to an upper layer (i.e., the owner of `CompositeObject`).

If exposing lifetime to an upper layer is not a desired option, you shouldn't store references in the first place.

Reply:
> 
> 
> > Responsibility of proper handling life-time is responsibility of `CompositeObject` and developer that written it
> > Bypassing life-time from "low-level" structures to "high-level" structures (logically) is violation of layers responsibilities
> 
> Actually, by storing references in `CompositeObject`, you are _explicitly_ delegating the responsibility of managing the lifetime of the pointee to an upper layer (i.e., the owner of `CompositeObject`).
> 
> If exposing lifetime to an upper layer is not a desired option, you shouldn't store references in the first place.


Instead I need to use ugly hacks like `Borrow` or `BorrowMut` which even do not provide life-time ?
Seems like by such strategy language are enforing users to user `Borrow` and `BorrowMut` and all goodice of Borrow Checker disappear  (

Reply:
Why are `Borrow`/`BorrowMut` relevant to any of this?

The problem here is one of fundamental soundness. If a struct stores a reference (directly, without `Box`ing), then all code which ever creates, manipulates or destroys instances of that struct _must_ be aware of that fact, or Rust the language (mostly the borrow checker in this case) can no longer guarantee "no UB in safe code". In particular, that means client code will fail to compile in some cases because of those stored references, and we want those compiler errors to make sense no matter how many layers of libraries there are in between.

When you really want to encapsulate a lifetime / make it an implementation detail clients don't need to care about, you simply need to make it live an arbitrarily long time by allocating, a.k.a. boxing it. That's true in every programming language; most of the others either autobox things by default or just let UB happen wherever. And yes, it's often a better design to pay the boxing cost to simplify things for your clients. But this whole proposal is (presumably) about cases where that boxing cost is unacceptable.

Reply:
> 
> 
> Why are `Borrow`/`BorrowMut` relevant to any of this?
> 
> The problem here is one of fundamental soundness. If a struct stores a reference (directly, without `Box`ing), then all code which ever creates, manipulates or destroys instances of that struct _must_ be aware of that fact, or Rust the language (mostly the borrow checker in this case) can no longer guarantee "no UB in safe code". In particular, that means client code will fail to compile in some cases because of those stored references, and we want those compiler errors to make sense no matter how many layers of libraries there are in between.

Okay, agree about this statement, what if just as was suggested in this topic to do something like this:
```rust
struct CompositeObject<...'_> {
    obj0: &SomeType, # Deduce life-time for this reference
    obj1: &SomeType, # Deduce life-time for this reference
}
```
and generates the same code as I suggest ?

> 
> 
> When you really want to encapsulate a lifetime / make it an implementation detail clients don't need to care about, you simply need to make it live an arbitrarily long time by allocating, a.k.a. boxing it. That's true in every programming language; most of the others either autobox things by default or just let UB happen wherever. And yes, while "SOLID" is pretty C++-specific and not directly applicable to most projects, it's often a better design to pay the boxing cost to simplify things for your clients. But this whole proposal is (presumably) about cases where that boxing cost is unacceptable, and therefore encapsulation is simply not an option.

`SOLID` is not C++-specific, it could be applied to anything that work with objects (structs, classes and so on), in AST all parts is some kind of object ... Anyway it is a holy-war and does not make sense to disscuss it, please take a look what I suggested above, will it be appropriate ? Syntax `...'_`

Reply:
What happens if `CompositeObject` is not just `CompositeObject<'anon>` but `CompositeObject<'anon0, 'anon1>`? How can one specify the lifetime of which variable relates to which part of the lifetime?

Reply:
@pickfire 
It will work as I described previously:
```rust
struct CompositeObject<...'_> {
    obj0: &SomeType, # Deduce life-time for this reference
    obj1: &SomeType, # Deduce life-time for this reference
}
```
to
```rust
struct CompositeObject<'anon0, 'anon1> {
    obj0: &'anon0 SomeType, # Deduce life-time for this reference
    obj1: &'anon1 SomeType, # Deduce life-time for this reference
}
```

I think that may this syntax would be more useful:
```rust
struct CompositeObject& {
    obj0: &SomeType, # Deduce life-time for this reference
    obj1: &SomeType, # Deduce life-time for this reference
}
```
and the compiler inference all life-times:
```rust
struct CompositeObject<'anon0, 'anon1> {
    obj0: &'anon0 SomeType, # Deduce life-time for this reference
    obj1: &'anon1 SomeType, # Deduce life-time for this reference
}
```
We would have goodness from all point of view ... User can see that `CompositeObject&` use some references inside, but at the same time user do not have to specify all life-times manually

Reply:
> We would have goodness from all point of view ... User can see that CompositeObject& use some references inside, but at the same time user do not have to specify all life-times manually

I understand that users can see some references, what I meant is what if the user wants to see all the references? Like checking if he/she could merge lifetime or something like that.

Reply:
> 
> 
> > We would have goodness from all point of view ... User can see that CompositeObject& use some references inside, but at the same time user do not have to specify all life-times manually
> 
> I understand that users can see some references, what I meant is what if the user wants to see all the references? Like checking if he/she could merge lifetime or something like that.

Then author should specify directly life-time, but if user should not know about life-time at all, user will see just `CompositeObject&` - which means that class could contain some references, be attention !!

Reply:
class?

Reply:
>  but if user should not know about life-time at all

That can't ever be a thing. Lifetimes need to always be publicly known as they affect usage.

Reply:
> 
> 
> > but if user should not know about life-time at all
> 
> That can't ever be a thing. Lifetimes need to always be publicly known as they affect usage.

It does not make sense !! You just want to see life-time, but there is no answer what you are doing with it ?

My answer - nothing, you just need to know that struct contains some references (from user point of view), and if there are some not trivial cases like this:
```rust
struct CompositeObject<'a, 'b: 'a> {
    obj0: &'a SomeType, # Deduce life-time for this reference
    obj1: &'b SomeType, # Deduce life-time for this reference
}
```
Developer still will not have other way, just to declare life-times

Reply:
Ah, now that is more understandable, so it means some cases are not covered. @redradist I think you should write that down in both the example (special case not supported) and in the drawbacks.

Reply:
Will the generated lifetime be in the docs or just used internally? If that is the case, maybe there is a term for it like "generated internally" or "syntactic sugar"?

Reply:
I do not think that generated life-time should be in docs due to that it does not have name ...
I think it should have just description that this structure has use references

Reply:
Yes, the lifetime needs to be in the docs. It is necessary for the user to know about the lifetime as it massively affects the usage of the struct. You need to design your code in a way that is compatible with the struct's lifetime semantics. Hiding those is incredibly dangerous in that it might mislead the developer into designing code that might not compile at all. Hiding lifetimes has already been deprecated, you are supposed to use at least `'_` nowadays, as it clarifies that a lifetime relationship is going on. So the absolute minimum that might make this RFC acceptable would be `struct Foo<'_>`. `<'_ ...>` probably is not working either, as that either just means the same thing as `<'_>` or it is incredibly fragile.

Reply:
> 
> 
> Yes, the lifetime needs to be in the docs. It is necessary for the user to know about the lifetime as it massively affects the usage of the struct. You need to design your code in a way that is compatible with the struct's lifetime semantics. Hiding those is incredibly dangerous in that it might mislead the developer into designing code that might not compile at all. Hiding lifetimes has already been deprecated, you are supposed to use at least `'_` nowadays, as it clarifies that a lifetime relationship is going on. So the absolute minimum that might make this RFC acceptable would be `struct Foo<'_>`. `<'_ ...>` probably is not working either, as that either just means the same thing as `<'_>` or it is incredibly fragile.

Take a look at examples above with syntax like this:
```rust
struct CompositeObject& {
    obj0: &SomeType, # Deduce life-time for this reference
    obj1: &SomeType, # Deduce life-time for this reference
}
```
or even just notification that in future structure will use references:
```rust
struct CompositeObject& { // No references inside, but may will be in future
    obj0: SomeType,
    obj1: SomeType,
}
```

Reply:
I believe at this point the RFC is basically lifetime elision in type declarations with `&` instead of `'_` (which imo there's no reason to introduce new syntax when we already have `'_` for this. And additionally it introduces automatic elision for `'static` which is probably a bit more controversial (and should probably be a separate RFC afterwards).

Reply:
> 
> 
> I believe at this point the RFC is basically lifetime elision in type declarations with `&` instead of `'_` (which imo there's no reason to introduce new syntax when we already have `'_` for this. And additionally it introduces automatic elision for `'static` which is probably a bit more controversial (and should probably be a separate RFC afterwards).

There is on big difference `'_` works and show only that in structure used only one reference with one life-time, but `CompositeObject&` shows that in structure used some references without explicitly declared life-time and number of fields could be different: 2, 3 ... whatever

Reply:
It doesn't, you can use a single lifetime but many references. Having more than one lifetime in a type's definition is only useful for disambiguating them for the user, but you can't do that with & anyway.

Reply:
> 
> 
> It doesn't, you can use a single lifetime but many references. Having more than one lifetime in a type's definition is only useful for disambiguating them for the user, but you can't do that with & anyway.

Okay, will it be possible to use `'_` life-time without reference inside ?
```rust
struct CompositeObject<'_> { // No references inside, but may will be in future
    obj0: SomeType,
    obj1: SomeType,
}
```